api/client/credentials.go

```diff
@@ -394,3 +400,193 @@ func configureTLS(c *tls.Config) *tls.Config {
 
 	return tlsConfig
 }
+
+// DynamicIdentityFileCreds allows a changing identity file to be used as the
+// source of authentication for Client. It does not automatically watch the
+// identity file or reload on an interval, this is left as an exercise for the
+// consumer.
+type DynamicIdentityFileCreds struct {
+	// mu protects the fields that may change if the underlying identity file
+	// is reloaded.
+	mu            sync.RWMutex
+	tlsCert       *tls.Certificate
+	tlsRootCAs    *x509.CertPool
+	sshCert       *ssh.Certificate
+	sshKey        crypto.Signer
+	sshKnownHosts []ssh.PublicKey
+	sshUser       string
+
+	// Path is the path to the identity file to load and reload.
+	Path string
+}
+
+// NewDynamicIdentityFileCreds returns a DynamicIdentityFileCreds which has
+// been initially loaded and is ready for use.
+func NewDynamicIdentityFileCreds(path string) (*DynamicIdentityFileCreds, error) {
+	d := &DynamicIdentityFileCreds{
+		Path: path,
+	}
+	if err := d.Reload(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return d, nil
+}
+
+// Reload causes the identity file to be re-read from the disk. It will return
+// an error if loading the credentials fails.
+func (d *DynamicIdentityFileCreds) Reload() error {
+	id, err := identityfile.ReadFile(d.Path)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// This section is essentially id.TLSConfig()
+	cert, err := keys.X509KeyPair(id.Certs.TLS, id.PrivateKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	pool := x509.NewCertPool()
+	for _, caCerts := range id.CACerts.TLS {
+		if !pool.AppendCertsFromPEM(caCerts) {
+			return trace.BadParameter("invalid CA cert PEM")
+		}
+	}
```

Comment:
nit: just personal preference. I would prefer calling `id.TLSConfig` and taking cert and pool from there or maybe saving the `identityTLSConfig` and using that later instead of "manually"  doing it here.

Reply:
I think I'd prefer to keep it as is. The reason for this is that `tls.Config` is a poor way of passing just a key/cert/CAs around - it has other fields that the underlying implementation could start using instead of the `Certificates` or `RootCAs` fields and I can easily see a future refactor to underlying parts breaking the `DynamicIdentityFileCreds` - the `.CACerts` and `.Certs` fields are much stricter and it's much more likely to be clearer that changing this risks breaking an upstream.

I have a somewhat long-term plan to try and convince folks that we should stop returning `tls.Config` and `ssh.ClientConfig` from the `client.Credentials` type. These configs are for more than just carrying certs/keys around and it's creating a conflict of responsibility for what the Credential should do and what the Client should do.